e2e node: Test probes during pod termination

hshiina opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:
These test cases look to intend to verify probe behaviors while a pod is being terminated:
https://github.com/kubernetes/kubernetes/blob/7ab8a2b92f675bb81715bfca7085b0c4d0a284c9/test/e2e_node/container_lifecycle_test.go#L756

In `SyncTerminatingPod()`, liveness probes are stopped before killing a pod while readiness probes are stopped after killing the pod:
https://github.com/kubernetes/kubernetes/blob/c9a61afdb7266b93b08da0d4ca4f334f62c3e7e2/pkg/kubelet/kubelet.go#L2024-L2038

The test cases are trying to test this code path with a failing liveness probe. However, when a liveness probe fails, the container is stopped at `SyncPod()` before the pod gets into termination phase even if `RestartPolicy` is `Never`:
https://github.com/kubernetes/kubernetes/blob/c9a61afdb7266b93b08da0d4ca4f334f62c3e7e2/pkg/kubelet/kubelet.go#L1963
https://github.com/kubernetes/kubernetes/blob/c9a61afdb7266b93b08da0d4ca4f334f62c3e7e2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L996-L999
So, both liveness probes and readiness probes can run while the preStop hook is running. The test case to see if the liveness probe is stopped passes just because the init container gets into `CrashLoopBackoff`. If the init container doesn’t get into `CrashLoopBackoff`, this test fails (#127312).

In order to run the preStop hook in `SyncTerminatingPod()`, this PR adds test cases to terminate a pod with  `ActiveDeadlineSeconds`.

After this is fixed, there remains a problem. Due to #124648, all probes are stopped before killing a pod if pod termination is triggered by kubelet such as active deadline and eviction. We need to fix this issue.

Alternatively, we can test this code path with deleting a pod. This PR also adds test cases to delete a pod with probes.

By the way, liveness probes for init containers are not stopped before killing the pod when the pod is terminated by deletion because `StopLivenessAndStartup()` stops only probes of regular containers:
https://github.com/kubernetes/kubernetes/blob/5ecd83f1e688e148a4ba7369392d7a58e212e1f6/pkg/kubelet/prober/prober_manager.go#L232

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hshiina committed:
e2e node: Test probes during pod termination

In order to test probes during pod termination, this fix uses
ActiveDeadlineSeconds to trigger the pod termination. A failing
liveness probe is not available for this purpose because a container
is stopped before pod termination.

hshiina commented:
/remove-kind bug
/kind cleanup

aojea commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -753,7 +753,7 @@ var _ = SIGDescribe(framework.WithNodeConformance(), "Containers Lifecycle", fun
 	})
 
-	ginkgo.When("a pod is terminating because its liveness probe fails", func() {
+	ginkgo.When("a pod is terminating because ActiveDeadlineSeconds expires", func() {
```
why do we modify an existing testcase instead of adding a new one?

We need to assert existing behavior does not change

bart0sh replied:
@hshiina please, address above review comment, thanks!

hshiina replied:
Sorry. I added new tests instead of modifying.

matthyx commented:
/ok-to-test

hshiina committed:
e2e node: Test probes during pod deletion

In order to test probes during pod termination, this fix add test cases
to delete a pod with probes to trigger the pod termination.

hshiina committed:
Restore removed test cases

hshiina commented:
/remove-kind flake

AnishShah commented:
/assign @AnishShah

bart0sh commented:
/triage accepted
/priority imprtant-longterm

bart0sh commented:
/priority important-longterm

HirazawaUi commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -907,6 +907,330 @@ var _ = SIGDescribe(framework.WithNodeConformance(), "Containers Lifecycle", fun
+							Name:  regular1,
+							Image: imageutils.GetE2EImage(imageutils.BusyBox),
+							Command: ExecCommand(regular1, execCommand{
+								Delay:              100,
```
Does this mean the pod will stay in the Running state for 100 seconds? This test takes way too long.

HirazawaUi commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -907,6 +907,330 @@ var _ = SIGDescribe(framework.WithNodeConformance(), "Containers Lifecycle", fun
+			framework.ExpectNoError(err)
+
+			// Sleep expecting thet the preStop wiil complete and the container will get into TerminationSeconds window.
+			time.Sleep(25 * time.Second)
```
I think waiting for 25 seconds here is unnecessary. We can synchronize the reduction of the required time through TerminationSeconds and the PreStop hook, so that a single test case can run at a faster speed.

HirazawaUi reviewed (commented):
This PR adds two new triggering conditions to the existing test cases. However, running them takes too much time. For the "ActiveDeadlineSeconds expires" condition, we even need 110+ seconds to complete a single test case. I believe this is unnecessary in practice. You should try to minimize the waiting time to make the tests run faster :)

hshiina reviewed (commented):
I will revise this PR with shortening duration after #129355 or #127865 is merged.

/hold

hshiina commented:
This is not related to this PR. I just want to see if how this job currently works.
/test pull-kubernetes-node-kubelet-serial-podresources

hshiina commented:
This is not related to this PR. I just want to see how this job currently works.
/test pull-kubernetes-e2e-gce-serial

hshiina commented:
/test pull-kubernetes-e2e-gce-serial

hshiina commented:
/test pull-kubernetes-e2e-gce-serial

hshiina commented:
/test pull-kubernetes-e2e-gce-serial-canary

hshiina commented:
/close
